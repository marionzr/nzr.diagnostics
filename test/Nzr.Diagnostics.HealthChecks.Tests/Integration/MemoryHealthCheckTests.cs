using FluentAssertions;
using Microsoft.Extensions.Diagnostics.HealthChecks;
using Microsoft.Extensions.Logging.Abstractions;
using Microsoft.Extensions.Options;
using NSubstitute;
using Snapshooter.Xunit;
namespace Nzr.Diagnostics.HealthChecks.Tests.Integration;

public class MemoryHealthCheckTests
{
    [Fact]
    public async Task MemoryHealthCheck_ShouldReturnHealthy_WhenMemoryIsNormal()
    {
        // Arrange

        // Setting the thresholds to the maximum to avoid flaky tests
        var optionsMonitor = CreateOptions(int.MaxValue - 1, int.MaxValue, int.MaxValue - 1, int.MaxValue);
        var healthCheck = new MemoryHealthCheck(optionsMonitor, new NullLogger<MemoryHealthCheck>());

        // Act

        var result = await healthCheck.CheckHealthAsync(new HealthCheckContext(), CancellationToken.None);

        // Assert

        result.Should().MatchSnapshot(options => options
            .IgnoreFields(nameof(HealthCheckResult.Data))
            .IgnoreField(nameof(HealthCheckResult.Description)));
    }

    private static IOptionsMonitor<MemoryHealthCheckOptions> CreateOptions(
        long warningThreshold,
        long criticalThreshold,
        long workingSetWarningThreshold,
        long workingSetCriticalThreshold)
    {
        var options = new MemoryHealthCheckOptions
        {
            WarningThreshold = warningThreshold,
            CriticalThreshold = criticalThreshold,
            WorkingSetWarningThreshold = workingSetWarningThreshold,
            WorkingSetCriticalThreshold = workingSetCriticalThreshold
        };

        var optionsMonitor = Substitute.For<IOptionsMonitor<MemoryHealthCheckOptions>>();
        optionsMonitor.CurrentValue.Returns(options);

        return optionsMonitor;
    }
}
