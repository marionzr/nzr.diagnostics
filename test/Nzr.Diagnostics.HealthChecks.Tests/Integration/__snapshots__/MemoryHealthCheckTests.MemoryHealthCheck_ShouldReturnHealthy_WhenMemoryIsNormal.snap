{
  "Data": {
    "AllocatedBytes": 3479424,
    "WorkingSet": 61460480,
    "PrivateMemorySize": 18534400,
    "Gen0Collections": 0,
    "Gen1Collections": 0,
    "Gen2Collections": 0,
    "HeapSize": 0,
    "CommittedMemory": 0,
    "FragmentedMemory": 0,
    "MemoryLoadPercentage": 0.0,
    "LargeObjectHeapSize": 0
  },
  "Description": "Memory usage is within normal range: Allocated: 4MB, Working Set: 59MB",
  "Exception": null,
  "Status": "Healthy"
}
